sql/core/src/test/scala/org/apache/spark/sql/FileBasedDataSourceSuite.scala

```diff
@@ -72,6 +72,29 @@ class FileBasedDataSourceSuite extends QueryTest with SharedSQLContext {
     }
   }
 
+  // Text and Parquet format does not allow wrting data frame with empty schema.
+  Seq("parquet", "text").foreach { format =>
+    test(s"SPARK-23372 writing empty dataframe should produce AnalysisException - $format") {
+      withTempPath { outputPath =>
+        intercept[AnalysisException] {
+          spark.emptyDataFrame.write.format(format).save(outputPath.toString)
+        }
+      }
+    }
+  }
+
+  // Formats excluding text and parquet allow writing empty data frames to files.
```

Comment:
I have added this test to show the current behaviour of reading the empty data frame.  For formats like orc, json, csv , we succeed to write an empty data frame but get the error while reading while inferring the schema ? Is this the right behaviour ?